src/ParquetMsg.chpl

```diff
@@ -705,6 +725,102 @@ module ParquetMsg {
     }
   }
 
+  inline proc getReaderIdx(fileNum: int, rgNum: int) {
+    // we can assume there won't be more than 1000 RGs in a file
+    return (fileNum*1000) + rgNum;
+  }
+
+  proc getRowGroupNums(ref distFiles, ref numRowGroups) {
+    coforall loc in distFiles.targetLocales() with (ref numRowGroups) do on loc {
+      var locFiles: [distFiles.localSubdomain()] string = distFiles[distFiles.localSubdomain()];
+      for i in locFiles.domain {
+        c_openFile(locFiles[i].localize().c_str(), getReaderIdx(i,0));
+        numRowGroups[i] = c_getNumRowGroups(getReaderIdx(i,0));
+        for j in 2..numRowGroups[i] {
+          c_openFile(locFiles[i].localize().c_str(), getReaderIdx(i,j-1));
+        }
+      }
+    }
+    var maxRowGroups = 0;
+    for val in numRowGroups do if maxRowGroups < val then maxRowGroups = val;
+    return maxRowGroups;
+  }
+  
+  proc fillSegmentsAndPersistData(ref distFiles, ref entrySeg, ref externalData, ref containsNulls, ref valsRead, dsetname, sizes, len, numRowGroups, ref bytesPerRG, ref startIdxs) throws {
+    var (subdoms, length) = getSubdomains(sizes);
+    coforall loc in distFiles.targetLocales() with (ref externalData, ref valsRead, ref bytesPerRG) do on loc {
+      var locFiles: [distFiles.localSubdomain()] string = distFiles[distFiles.localSubdomain()];
+      var locSubdoms = subdoms;
+
+      for i in locFiles.domain {
+        var fname = locFiles[i];
+        var locDsetname = dsetname;
+        for rg in 0..#numRowGroups[i] {
+          c_createRowGroupReader(rg, getReaderIdx(i,rg));
+          c_createColumnReader(locDsetname.localize().c_str(), getReaderIdx(i,rg));
+        }
+      }
+
+      var errs: [locFiles.domain] (bool, parquetErrorMsg) = [0..#locFiles.domain] (false, new parquetErrorMsg());
+      forall i in locFiles.domain {
+        var fname = locFiles[i];
+        var locDsetname = dsetname;
+        var startIdx = locSubdoms[i].low;
+        for rg in 0..#numRowGroups[i] {
+          var totalBytes = 0;
+          startIdxs[i][rg] = startIdx;
+
+          var numRead = 0;
+
+          if c_readParquetColumnChunks(fname.localize().c_str(), 8192, len, getReaderIdx(i,rg), c_ptrTo(numRead), c_ptrTo(externalData[i][rg]), c_ptrTo(containsNulls[i][rg]), c_ptrTo(errs[i][1].errMsg)) == ARROWERROR {
+            errs[i] = (true, errs[i][1]);
+          }
+          var tmp: [startIdx..#numRead] int;
+          forall (id, j) in zip(0..#numRead, startIdx..#numRead) with (+ reduce totalBytes) {
+            ref curr = (externalData[i][rg]: c_ptr(MyByteArray))[id];
+            tmp[j] = curr.len + 1; // this was only change
+            totalBytes += curr.len+1;
+          }
+          entrySeg.a[startIdx..#numRead] = tmp;
```

Comment:
I'm struggling to understand exactly how this bit works

Reply:
The external data holds the Parquet `ByteArray`s that actually store the data and we need to get all the bytes from that data to create our Chapel array. So, we do that by getting the chunk of this columns data (`externalData[i][rg]`) and then cast that to a byte array (it's a C++ class, so we have to have a C wrapper to it, which is why `MyByteArray` isn't the real Arrow type, but it's same amount of memory) and then we get the actual individual string with the `[id]` index. From there, we can get the total number of bytes by adding each strings length and then we get the segmented lengths of the actual strings into our seg array here too with the `tmp[j] = curr.len+1;`. The `+1` is because the Arrow strings are not null-terminated, so we need to have 1 more character in our Arkouda array per string.

Reply:
ah! I think the biggest misunderstanding here was me being real dumb and not reading the name of `entrySeg` lol. I thought that was the vals 🤦‍♀️  this makes sense now, thanks ben!